examples/purchase.rs

```diff
@@ -0,0 +1,145 @@
+// This example demonstrates how to deal with messages and callback queries
+// within a single dialogue.
+//
+// # Example
+// ```
+// - /start
+// - Let's start! What's your full name?
+// - John Doe
+// - Select a product:
+//   [Apple, Banana, Orange, Potato]
+// - <A user selects "Banana">
+// - John Doe, product 'Banana' has been purchased successfully!
+// ```
+
+use teloxide::{
+    dispatching::dialogue::{self, GetChatId, InMemStorage},
+    prelude::*,
+    types::{InlineKeyboardButton, InlineKeyboardMarkup},
+    utils::command::BotCommands,
+};
+
+type MyDialogue = Dialogue<State, InMemStorage<State>>;
+type HandlerResult = Result<(), Box<dyn std::error::Error + Send + Sync>>;
+
+#[derive(Clone)]
+pub enum State {
+    Start,
+    ReceiveFullName,
+    ProductChosen { full_name: String },
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self::Start
+    }
+}
+
+#[derive(BotCommands, Clone)]
+#[command(rename = "lowercase", description = "These commands are supported:")]
+enum Command {
+    #[command(description = "display this text.")]
+    Help,
+    #[command(description = "start the purchase procedure.")]
+    Start,
+}
+
+#[tokio::main]
+async fn main() {
+    pretty_env_logger::init();
+    log::info!("Starting dialogue_bot...");
+
+    let bot = Bot::from_env().auto_send();
+
+    Dispatcher::builder(
+        bot,
+        dialogue::enter::<Update, InMemStorage<State>, State, _>()
+            .branch(
+                Update::filter_message()
+                    .chain(teloxide::handler![State::ReceiveFullName].endpoint(receive_full_name)),
+            )
+            .branch(
+                Update::filter_callback_query().chain(
+                    teloxide::handler![State::ProductChosen { full_name }]
+                        .endpoint(receive_product_selection),
+                ),
+            )
+            .branch(Update::filter_message().filter_command::<Command>().endpoint(handle_command))
+            .branch(Update::filter_message().endpoint(invalid_state)),
```

Comment:
Shouldn't all message branches use the same filter?

Reply:
I think this wouldn't make the code simpler.

![Screenshot from 2022-04-23 22-38-02](https://user-images.githubusercontent.com/40539574/164915198-4637e0b4-d566-45f0-aad2-3d4b3b92bb38.png)

Reply:
`invalid_state` can reuse the same `Update::filter_message()`, IMO it's less confusing when all updates of a given type are processed in the same subtree.

Reply:
It couldn't, because then we cannot handle callback queries because the first branch would always work.